engine/notifications.js

```diff
@@ -0,0 +1,283 @@
+        return Promise.all(features.map(slug => [
+          redis.srem(`${slug}-notifications`, deviceId),
+          redis.srem(`${deviceId}-notifications`, slug)])
+        );
+      }
+      // user decided to unregister from `all`
+      if (features.indexOf('all') >= 0) {
+        return redis.srem('all-notifications', deviceId)
+        .then(() => redis.srem(`${deviceId}-notifications`, 'all'));
+      }
+      // unregister from `all` and register to negative of `features`
+      return redis.srem('all-notifications', deviceId)
+      .then(() => redis.srem(`${deviceId}-notifications`, 'all'))
+      // register to everything except of features
+      .then(() => redis.get('status'))
+      .then(status => {
+        let allFeatures = ['new'];
+        allFeatures = allFeatures.concat(Object.keys(JSON.parse(status)));
+        return Promise.all(allFeatures.map(slug => {
+          if (features.indexOf(slug) >= 0) {
+            return Promise.resolve();
+          }
+          return redis.sadd(`${deviceId}-notifications`, slug)
+          .then(() => redis.sadd(`${slug}-notifications`, deviceId));
+        }));
+      });
+    });
+  })
+  .then(() => getRegisteredFeatures(deviceId))
+  .catch(err => {
+    if (err.message !== 'Not Found') {
+      throw err;
+    }
+    return [];
+  })
+  .then(regFeatures => {
+    if (!doNotConfirm && endpoint) {
+      return sendConfirmation(endpoint, key, authSecret, deviceId, regFeatures)
+      .then(() => regFeatures);
+    }
+    return regFeatures;
+  });
+}
+
+// Registers to all notifications including new ones
+// Removes registrations to individual features
+function registerToAll(deviceId) {
+  return getRegisteredFeatures(deviceId)
+  .then(features => unregister(deviceId, features, true))
+  .then(() => Promise.all([
+    redis.sadd('all-notifications', deviceId),
+    redis.sadd(`${deviceId}-notifications`, 'all')]));
+}
+
+// registers to receive notifications
+// required info:
+// * deviceId
+// * list of feature slugs to register to
+//   special features are `new` and `all`
+//   * new - do notify about new features
+//   * all - all features including new, registering to all will remove
+//     all individual registrations
+// * endpoint [optional]
+// one can register to a feature without providing an endpoint
+// if it was already saved:
+// * key
+function register(deviceId, features, endpoint, key, authSecret) {
+  if (!features || features.length === 0) {
+    return Promise.reject(new Error('No features provided'));
+  }
+
+  return Promise.resolve()
+  .then(() => {
+    if (endpoint) {
+      const device = {
+        id: deviceId,
+        endpoint,
+        key: key || '',
+        authSecret: authSecret || '',
+      };
+      return redis.hmset(`device-${deviceId}`, device);
+    }
+  })
+  .then(() => redis.hgetall(`device-${deviceId}`))
+  .then(device => {
+    if (!device || !device.endpoint) {
+      throw new Error('No endpoint provided');
```

Comment:
Can you update the exception message then? (something like 'No endpoint provided and no endpoint was found in DB').